design/local-node-agent.md

```diff
@@ -0,0 +1,147 @@
+When the agent starts up, it will immediately copy the Rook Flexvolume driver from its image to the `volume-plugin-dir` on its host node.
+In Kubernetes 1.8+, the [dynamic Flexvolume plugin discovery](https://github.com/kubernetes/community/pull/833) will find and initialize our driver, but in older versions of Kubernetes a manual restart of the Kubelet will be required.
+
+After the driver has been deployed, the Rook agent will create a Unix domain socket that will serve as the communications channel between the agent and the Flexvolume driver.
+The driver will initiate a handshake to the agent over the socket during its [`Init` function](https://github.com/kubernetes/community/blob/master/contributors/devel/flexvolume.md#init) when it is executed after being discovered by the Controller manager.
+The agent will wait for the driver to complete the handshake before moving on and being ready to accept Flexvolume storage requests.
+
+Note that the Rook operator will continue to run its existing [dynamic volume provisioner](https://github.com/rook/rook/tree/master/pkg/operator/provisioner) to provision and delete persistent volumes as needed.
+
+### **Flexvolume**
+#### Control Flow Overview
+The Rook Flexvolume driver will be very lightweight, simply implementing the [required interface](https://github.com/kubernetes/community/blob/master/contributors/devel/flexvolume.md#driver-invocation-model)
+and then offloading the storage provider work to the Rook agent over the Unix domain socket.
+Below is a simplified description of the control flow for providing block storage ready to be consumed by a pod:
+
+1. Rook operator is running and a Rook agent pod is running on every node in the cluster.  The Rook Flexvolume driver has been deployed to the `volume-plugin-dir` on each node.
+1. A user creates a Persistent Volume Claim (PVC) specifying a storageclass that uses the `rook.io/block` [provisioner](https://github.com/kubernetes-incubator/external-storage/blob/e6e64ad1a431fea37f723882f36251f8d2fe4247/lib/controller/volume.go#L29)
+1. `Provision()` is called on the operator's provisioner to create a block image in the cluster.  At this point, the Provision phase is complete and the PVC/PV are considered `Bound` together.
+1. Once a pod has been created that consumes the PVC, `Attach()` is called by the attach/detach controller on the Rook Flexvolume on the Controller manager node, which calls into its local Rook agent via the Unix domain socket.
+The agent then uses the Kubernetes API to create a volume attach CRD in the namespace where the Rook cluster providing the storage resides, declaring the intent that an attach needs to happen between a volume and a node.
+   * The attach/detach controller waits on the driver which in turn blocks/waits for the attach to complete on the remote node.
+1. All Rook agents are watching for volume attach CRD events, filtering by a `node=<node name>` label on the CRD instance.
```

Comment:
The filter may/may not happen. Depends on the https://github.com/kubernetes/kubernetes/issues/49424, which seems not to be slated for 1.8. I'll try to experiment with a different way to watch objects instead of the more generic one.

Reply:
The intent is that eventually the optimal filter will happen (whether or not it happens right away).  I have linked to the upstream tracking issue elsewhere in the document, so I think we're OK.